lib/vm/src/table.rs

```diff
@@ -102,11 +104,58 @@ pub trait Table: fmt::Debug + Send + Sync {
     }
 }
 
+/// A reference stored in a table. Can be either an externref or a funcref.
+#[derive(Debug, Clone)]
+pub enum TableReference {
+    // TODO: implement extern refs
+    /// Opaque pointer to arbitrary host data.
+    ExternRef(VMExternRef),
+    /// Pointer to function: contains enough information to call it.
+    FuncRef(VMFuncRef),
+}
+
+impl From<TableReference> for TableElement {
+    fn from(other: TableReference) -> Self {
+        match other {
+            TableReference::ExternRef(extern_ref) => Self { extern_ref },
+            TableReference::FuncRef(func_ref) => Self { func_ref },
+        }
+    }
+}
+
+// TODO: review what repr(C) means on unions, make sure this just acts like a usize.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub union TableElement {
+    pub(crate) extern_ref: VMExternRef,
+    pub(crate) func_ref: VMFuncRef,
+}
```

Comment:
I'm unable to see the difference between `TableReference` and `TableElement` (other than the internal representation).
We should probably simplify this.

Reply:
Well the intent is that their internal representation is different. `union` means the size of the entire thing is the size of biggest thing in it, so it's 1 word. Using an enum means you need extra space for the tag. We could use an `enum` and `usize` as a `void*` in some places but I don't know if I'd consider that a simplification.

What are you thinking here for simplifying it?